homeassistant/components/plaato/sensor.py

```diff
@@ -0,0 +1,142 @@
+"""Support for Plaato Airlock sensors."""
+
+import logging
+
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+
+from . import (
+    ATTR_ABV, ATTR_BATCH_VOLUME, ATTR_BPM, ATTR_CO2_VOLUME, ATTR_TEMP,
+    ATTR_TEMP_UNIT, ATTR_VOLUME_UNIT, DOMAIN as PLAATO_DOMAIN,
+    PLAATO_DEVICE_ATTRS, PLAATO_DEVICE_SENSORS, SENSOR as SENSOR_DOMAIN,
+    SENSOR_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_KEY = '{}.{}'.format(PLAATO_DOMAIN, SENSOR_DOMAIN)
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the Plaato sensor."""
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Plaato from a config entry."""
+    devices = {}
+
+    def get_device(device_id):
+        """Get a device."""
+        return hass.data[PLAATO_DOMAIN].get(device_id, False)
+
+    def get_device_sensors(device_id):
+        """Get device sensors."""
+        return hass.data[PLAATO_DOMAIN].get(device_id)\
+            .get(PLAATO_DEVICE_SENSORS)
+
+    async def _update_sensor(device_id):
+        """Update/Create the sensors."""
+        if device_id not in devices and get_device(device_id):
+            entities = []
+            sensors = get_device_sensors(device_id)
+
+            for sensor_type in sensors.items():
+                entities.append(PlaatoSensor(device_id, sensor_type))
+
+            devices[device_id] = entities
+
+            async_add_entities(entities, True)
+        else:
+            for entity in devices[device_id]:
+                entity.async_schedule_update_ha_state()
```

Comment:
Pass `True` so that it will call the `async_update` method and fetch the latest data.

```suggestion
                entity.async_schedule_update_ha_state(True)
```

Reply:
Alternative, since the data is just fetched from a dictionary stored in memory during update, you could have all properties fetch the data directly from that dictionary instead of storing it on the entity instance. In that case you can skip `True`

Reply:
What you are suggesting would mean that I should move things from `async_update` to the `state` property? Something like this?

```python
    @property
    def state(self):
        """Return the state of the sensor."""
         sensors = self.get_sensors()
        if sensors is False:
            _LOGGER.debug("Device with name %s has no sensors.", self.name)
            return 0

        if self._type == ATTR_ABV:
            return round(sensors.get(self._type), 2)
        elif self._type == ATTR_TEMP:
              return round(sensors.get(self._type), 1)
        elif self._type == ATTR_CO2_VOLUME:
            return round(sensors.get(self._type), 2)
        else:
            return sensors.get(self._type)
```

Reply:
yeah. Just know that you cannot do any I/O

Reply:
Done!

Reply:
It's not safe to call `entity.async_schedule_update_ha_state` directly in non polling platforms.

For polling platforms we can assume that the interval between polls is long enough so that all new entities will have been added to home assistant before the next update, via poll, is done and calls this method.

For non polling platforms we can't assume this. The next update for new entities might come immediately after the first one, before they have been added to home assistant. Then this call will error.

The safe approach in this case is to use our dispatch helper and send a signal to the entity to have it call `async_schedule_update_ha_state` from inside itself. The signal should be connected in `async_added_to_hass` which will guard from the error.

Reply:
Thank @MartinHjelmare I will look into this. You don't happen to have any examples where this behavior is used? So I can learn how to do it.

Reply:
Something like this:
https://github.com/home-assistant/home-assistant/blob/f382be4c15ef32f557ba6d40bb01c084c5cbc692/homeassistant/components/aftership/sensor.py#L145-L153

In our case we might not even need an extra method that does the update call we can connect `async_schedule_update_ha_state` directly. So replace `self.force_update` with `self.async_schedule_update_ha_state`.

Reply:
@MartinHjelmare Thank you. Should this change go into a new PR or can I push it here and it will be merged in?

Reply:
@MartinHjelmare PR with the changes you requested: https://github.com/home-assistant/home-assistant/pull/24627